src/app/clusters/on-off-server/on-off-server.cpp

```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
+            OOattributeValueList));
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        serialisedBytes.reduce_size(writer.GetLengthWritten());
+
+        return CHIP_NO_ERROR;
+    }
+
+    /// @brief Default EFS interaction when applying scene to the OnOff Cluster
+    /// @param endpoint target endpoint
+    /// @param cluster  target cluster
+    /// @param serialisedBytes Data from nvm
+    /// @param timeMs transition time in ms
+    /// @return CHIP_NO_ERROR if value as expected, CHIP_ERROR_INVALID_ARGUMENT otherwise
+    CHIP_ERROR ApplyScene(EndpointId endpoint, ClusterId cluster, const ByteSpan & serialisedBytes,
+                          scenes::TransitionTimeMs timeMs) override
+    {
+        app::DataModel::DecodableList<Scenes::Structs::AttributeValuePair::DecodableType> attributeValueList;
+        Scenes::Structs::AttributeValuePair::DecodableType decodePair;
+
+        TLV::TLVReader reader;
+        TLV::TLVType outer;
+
+        size_t attributeCount = 0;
+
+        VerifyOrReturnError(cluster == Id, CHIP_ERROR_INVALID_ARGUMENT);
+
+        reader.Init(serialisedBytes);
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+        ReturnErrorOnFailure(reader.EnterContainer(outer));
+        ReturnErrorOnFailure(reader.Next(
+            TLV::kTLVType_Array, TLV::ContextTag(app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList)));
+        ReturnErrorOnFailure(attributeValueList.Decode(reader));
+
+        ReturnErrorOnFailure(attributeValueList.ComputeSize(&attributeCount));
+        VerifyOrReturnError(attributeCount <= scenableAttributeCount, CHIP_ERROR_BUFFER_TOO_SMALL);
+        auto pair_iterator = attributeValueList.begin();
+
+        // Assumes the OnOff cluster has only one Attribute
+        pair_iterator.Next();
+
+        decodePair = pair_iterator.GetValue();
+
+        if (decodePair.attributeID.HasValue())
+        {
+            // If attribute ID was encoded, verify it is the proper ID for the OnOff attribute
+            VerifyOrReturnError(decodePair.attributeID.Value() == Attributes::OnOff::Id, CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        mSceneOnOffState = decodePair.attributeValue;
+
+        // Verify that the EFS was completely read
+        ReturnErrorOnFailure(pair_iterator.GetStatus());
+        ReturnErrorOnFailure(reader.ExitContainer(outer));
+
+        OnOffServer::Instance().scheduleTimerCallbackMs(sceneEventControl(endpoint), timeMs);
+
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    /**
+     * @brief Configures EnventControl callback when setting On Off through scenes callback
+     *
+     * @param[in] endpoint endpoint to start timer for
+     * @return EmberEventControl* configured event control
+     */
+    EmberEventControl * sceneEventControl(EndpointId endpoint)
+    {
+        EmberEventControl * controller = OnOffServer::Instance().getEventControl(endpoint);
```

Comment:
This is going to wipe out any existing timing going on for OnWithTimedOff, right?  Why is that OK?

Reply:
Isn't that the desired behavior? Say I change scene and user presses a button toggling onOff, it will override the change scene, and vice-versa, if user presses a button and change scenes happen in between that press and the end of the transition time. 

Do we want to queue different event every-time a new scene is invoke? Say if a user calls recall scene on scene1 with a transition time of 30s, and then changes his mind and calls recall scene2 with a transition of 10s, then the scene would transition to scene2, then scene1. Similar behavior could happen between user action and scene recall. Say user presses button, which starts a transition on-off -> 1, then recall scene2 which has on-off = 0 and a transition time of 0.

Scene2 makes on-off=0, then the button press happens and cancels user option. This doesn't seem like the desired behavior.

Reply:
Let me ask a more direct question:  If I do OnWithTimedOff, and then I change scene, should that cancel the timed off?  The timed off bits are generally meant to be used by sensor-based lighting setups, I believe, and I would have thought that you might want that even if you have scenes going on.  @Rob-Houtepen-Signify thoughts?

Reply:
The OnWithTimedOff related use-case is indeed about sensor based for automatic control, but allowing for manual overwrites. Originally, people where more thinking about On/Off commands as result of a manual override, but like you indicated, a Recall Scene command may also be the result of a manual override.

The spec text below the state diagram says:
Off1 : Any command which causes the OnOff attribute to be set to FALSE, e.g. Off, Toggle or OffWithEffect.
On2 : Any command which causes the OnOff attribute to be set to TRUE, e.g. On, Toggle or OnWithRecallGlobalScene.

So on a Recall Scene I would check whether the resulting state for the OnOff attribute is True or False, and treat this as an internal On of Off command. (Even if the OnOff attribute does not change, still treat it as internal On or Off command to have the effect of the manual override.)

Reply:
@Rob-Houtepen-Signify The real issue I am having is that this proposed code is reusing the timer that drives the OnTime/OffWaitTime updates to time the scene transition.

So for example, if we are in "Delayed Off" state and we get a scene change with OnOff = false, per spec it seems like it should leave us in that state and keep decrementing OffWaitTime.  But as far as I can tell it will instead stop updating OffWaitTime, and probably leave us stuck in "Delayed Off".  If a lot of time passes, the next OnWithTimedOff command, will start up the timer again but stay in "Delayed Off" instead of transitioning to "Timed On" as it should....

The actual application of the scene does in fact pretend to be an On or Off command, so that part is fine.

@lpbeliveau-silabs am I missing something in the analysis above?

Reply:
@bzbarsky-apple That is correct, we will indeed restart from where we were at once the next function is called in some cases. 
What I propose is to add:
`uint16_t onTime                     = 0;
uint16_t offWaitTime                = 0;  
OnOff::Attributes::OffWaitTime::Set(endpoint, &offWaitTime);
OnOff::Attributes::OnTime::Set(endpoint, &onTime);`
(In a proper styling )To the original sceneOnOffCallback() below.

Reply:
I don't think resetting OnTime and OffWaitTime on scene change would match the spec.

If the idea is that scene change is supposed to act as if a On or Off was sent, then per "1.5.8. State Description" in appclusters:

1. Changing scene to OnOff = false while in "Delayed Off" state should not interfere with the OffWaitTime countdown.
2. Changing scene to OnOff = true while in "Timed On" state should not interfere with the OnTime countdown.

Or at least that's my understanding of the spec and of @Rob-Houtepen-Signify's comments above.

Reply:
Current version now has a sceneHandlerEvenControl array for each endpoint